File: packages/heady-semantic-logic/src/core/dynamic-nibble.ts
/**
 * Dynamic Nibble Assignment System
 * Allows runtime selection of bit-depth (4, 8, 16, 32, 64, 128, 256, 4096 bits)
 * for adaptive precision based on task requirements
 */

export type NibbleBitDepth = 4 | 8 | 16 | 32 | 64 | 128 | 256 | 512 | 1024 | 2048 | 4096;

export interface DynamicNibbleConfig {
  bitDepth: NibbleBitDepth;
  name: string;
  decimalDigits: number;
  speedFactor: number;  // Relative to 64-bit standard
  memoryBytes: number;
}

/**
 * Predefined nibble configurations for common use cases
 */
export const NIBBLE_PRESETS: Record<string, DynamicNibbleConfig> = {
  'ultra_low_power': { bitDepth: 4, name: 'ULP_4bit', decimalDigits: 1, speedFactor: 10, memoryBytes: 0.5 },
  'iot_edge': { bitDepth: 8, name: 'IoT_8bit', decimalDigits: 2, speedFactor: 5, memoryBytes: 1 },
  'embedded': { bitDepth: 16, name: 'FP16', decimalDigits: 3, speedFactor: 2, memoryBytes: 2 },
  'standard': { bitDepth: 64, name: 'FP64', decimalDigits: 15, speedFactor: 1, memoryBytes: 8 },
  'high': { bitDepth: 128, name: 'Quad128', decimalDigits: 34, speedFactor: 0.02, memoryBytes: 16 },
  'octuple': { bitDepth: 256, name: 'Octuple256', decimalDigits: 71, speedFactor: 0.001, memoryBytes: 32 },
  'crypto_1024': { bitDepth: 1024, name: 'Crypto1K', decimalDigits: 308, speedFactor: 0.0001, memoryBytes: 128 },
  'crypto_2048': { bitDepth: 2048, name: 'Crypto2K', decimalDigits: 617, speedFactor: 0.00005, memoryBytes: 256 },
  'rsa_4096': { bitDepth: 4096, name: 'RSA4K', decimalDigits: 1233, speedFactor: 0.00001, memoryBytes: 512 }
};

/**
 * Dynamic Nibble Manager for runtime bit-depth selection
 */
export class DynamicNibbleManager {
  private activeConfig: DynamicNibbleConfig;
  private conversionCache = new Map<string, any>();

  constructor(config: DynamicNibbleConfig | string = 'standard') {
    this.activeConfig = typeof config === 'string' 
      ? NIBBLE_PRESETS[config] 
      : config;
  }

  getConfig(): DynamicNibbleConfig {
    return this.activeConfig;
  }

  /**
   * Change bit depth at runtime
   */
  switchBitDepth(newConfig: DynamicNibbleConfig | string): void {
    this.activeConfig = typeof newConfig === 'string'
      ? NIBBLE_PRESETS[newConfig]
      : newConfig;
    this.conversionCache.clear();
  }

  /**
   * Convert truth value to active bit depth
   */
  encode(value: number): bigint | number {
    const key = `encode_${value}_${this.activeConfig.bitDepth}`;
    if (this.conversionCache.has(key)) {
      return this.conversionCache.get(key);
    }

    const clamped = Math.max(0, Math.min(1, value));
    const maxValue = (BigInt(1) << BigInt(this.activeConfig.bitDepth)) - BigInt(1);
    const encoded = BigInt(Math.floor(clamped * Number(maxValue)));

    this.conversionCache.set(key, encoded);
    return encoded;
  }

  /**
   * Decode from active bit depth back to [0, 1]
   */
  decode(encoded: bigint | number): number {
    const key = `decode_${encoded}_${this.activeConfig.bitDepth}`;
    if (this.conversionCache.has(key)) {
      return this.conversionCache.get(key);
    }

    const maxValue = (BigInt(1) << BigInt(this.activeConfig.bitDepth)) - BigInt(1);
    const decoded = Number(BigInt(encoded)) / Number(maxValue);

    this.conversionCache.set(key, decoded);
    return decoded;
  }

  /**
   * Convert between different bit depths
   */
  transcode(value: bigint | number, fromBits: NibbleBitDepth, toBits: NibbleBitDepth): bigint {
    const fromMax = (BigInt(1) << BigInt(fromBits)) - BigInt(1);
    const toMax = (BigInt(1) << BigInt(toBits)) - BigInt(1);

    const normalized = Number(BigInt(value)) / Number(fromMax);
    return BigInt(Math.floor(normalized * Number(toMax)));
  }

  /**
   * Adaptive bit-depth selection based on task characteristics
   */
  static selectOptimalBitDepth(task: {
    realtime?: boolean;
    cryptographic?: boolean;
    financial?: boolean;
    iot?: boolean;
    accuracy_required?: 'low' | 'medium' | 'high' | 'extreme';
    iterations?: number;
  }): string {
    // Real-time tasks prioritize speed
    if (task.realtime) return 'standard';

    // IoT/edge devices need low power
    if (task.iot) return 'iot_edge';

    // Cryptographic operations need high precision
    if (task.cryptographic) return 'rsa_4096';

    // Financial calculations need balanced precision
    if (task.financial) return 'octuple';

    // Long-running simulations accumulate error
    if (task.iterations && task.iterations > 1e6) return 'octuple';

    // Accuracy-based selection
    switch (task.accuracy_required) {
      case 'low': return 'embedded';
      case 'medium': return 'standard';
      case 'high': return 'high';
      case 'extreme': return 'rsa_4096';
      default: return 'standard';
    }
  }

  /**
   * Get estimated performance characteristics
   */
  estimatePerformance(operations: number): {
    executionTimeMs: number;
    memoryUsageMB: number;
    powerConsumptionMw: number;
  } {
    const baseTimeMs = 0.001; // 1μs per operation at standard precision
    const executionTimeMs = (operations * baseTimeMs) / this.activeConfig.speedFactor;

    const memoryUsageMB = (operations * this.activeConfig.memoryBytes) / (1024 * 1024);

    // Power scales with memory bandwidth and computation time
    const powerConsumptionMw = (memoryUsageMB * 10) + (executionTimeMs * 0.5);

    return { executionTimeMs, memoryUsageMB, powerConsumptionMw };
  }

  toString(): string {
    return `DynamicNibble[${this.activeConfig.name}, ${this.activeConfig.bitDepth}bit, ${this.activeConfig.decimalDigits}digits]`;
  }
}

/**
 * Multi-resolution gate that operates on different bit depths
 */
export class MultiResolutionGate {
  private nibbleManager: DynamicNibbleManager;

  constructor(config: DynamicNibbleConfig | string = 'standard') {
    this.nibbleManager = new DynamicNibbleManager(config);
  }

  /**
   * AND gate with dynamic bit depth
   */
  AND(inputs: number[]): number {
    const encoded = inputs.map(v => this.nibbleManager.encode(v));

    // Min operation in encoded space
    const minEncoded = encoded.reduce((min, val) => 
      BigInt(val) < BigInt(min) ? val : min
    );

    return this.nibbleManager.decode(minEncoded);
  }

  /**
   * OR gate with dynamic bit depth
   */
  OR(inputs: number[]): number {
    const encoded = inputs.map(v => this.nibbleManager.encode(v));

    // Max operation in encoded space
    const maxEncoded = encoded.reduce((max, val) => 
      BigInt(val) > BigInt(max) ? val : max
    );

    return this.nibbleManager.decode(maxEncoded);
  }

  /**
   * Switch resolution at runtime
   */
  setResolution(config: string | DynamicNibbleConfig): void {
    this.nibbleManager.switchBitDepth(config);
  }

  getResolution(): DynamicNibbleConfig {
    return this.nibbleManager.getConfig();
  }
}

/**
 * Adaptive nibble pipeline that adjusts precision based on load
 */
export class AdaptiveNibblePipeline {
  private stages: Map<string, DynamicNibbleManager> = new Map();
  private performanceHistory: number[] = [];

  constructor() {
    // Initialize with different precision stages
    this.stages.set('intake', new DynamicNibbleManager('standard'));
    this.stages.set('processing', new DynamicNibbleManager('high'));
    this.stages.set('output', new DynamicNibbleManager('standard'));
  }

  /**
   * Auto-tune precision based on system load
   */
  autoTune(systemLoad: number): void {
    // High load (>80%): reduce precision for speed
    if (systemLoad > 0.8) {
      this.stages.get('processing')?.switchBitDepth('standard');
    } 
    // Low load (<30%): increase precision for accuracy
    else if (systemLoad < 0.3) {
      this.stages.get('processing')?.switchBitDepth('octuple');
    }
    // Medium load: balanced precision
    else {
      this.stages.get('processing')?.switchBitDepth('high');
    }
  }

  /**
   * Process value through adaptive pipeline
   */
  process(value: number, stage: string = 'processing'): number {
    const manager = this.stages.get(stage);
    if (!manager) throw new Error(`Unknown stage: ${stage}`);

    const encoded = manager.encode(value);
    return manager.decode(encoded);
  }

  /**
   * Get current pipeline configuration
   */
  getPipelineStatus(): Record<string, string> {
    const status: Record<string, string> = {};
    for (const [stage, manager] of this.stages) {
      status[stage] = manager.getConfig().name;
    }
    return status;
  }
}

File: packages/heady-semantic-logic-python/pqc/dynamic_nibble.py
"""
Dynamic Nibble Assignment System for Python
Runtime bit-depth selection (4, 8, 16, 32, 64, 128, 256, 4096 bits)
"""

from typing import Dict, Union, List
from dataclasses import dataclass

@dataclass
class DynamicNibbleConfig:
    bit_depth: int
    name: str
    decimal_digits: int
    speed_factor: float
    memory_bytes: float

NIBBLE_PRESETS: Dict[str, DynamicNibbleConfig] = {
    'ultra_low_power': DynamicNibbleConfig(4, 'ULP_4bit', 1, 10.0, 0.5),
    'iot_edge': DynamicNibbleConfig(8, 'IoT_8bit', 2, 5.0, 1),
    'embedded': DynamicNibbleConfig(16, 'FP16', 3, 2.0, 2),
    'standard': DynamicNibbleConfig(64, 'FP64', 15, 1.0, 8),
    'high': DynamicNibbleConfig(128, 'Quad128', 34, 0.02, 16),
    'octuple': DynamicNibbleConfig(256, 'Octuple256', 71, 0.001, 32),
    'crypto_1024': DynamicNibbleConfig(1024, 'Crypto1K', 308, 0.0001, 128),
    'rsa_4096': DynamicNibbleConfig(4096, 'RSA4K', 1233, 0.00001, 512)
}

class DynamicNibbleManager:
    """Runtime bit-depth manager"""

    def __init__(self, config: Union[str, DynamicNibbleConfig] = 'standard'):
        if isinstance(config, str):
            self.active_config = NIBBLE_PRESETS[config]
        else:
            self.active_config = config
        self.conversion_cache = {}

    def get_config(self) -> DynamicNibbleConfig:
        return self.active_config

    def switch_bit_depth(self, new_config: Union[str, DynamicNibbleConfig]):
        """Change bit depth at runtime"""
        if isinstance(new_config, str):
            self.active_config = NIBBLE_PRESETS[new_config]
        else:
            self.active_config = new_config
        self.conversion_cache.clear()

    def encode(self, value: float) -> int:
        """Convert [0,1] float to active bit depth"""
        clamped = max(0.0, min(1.0, value))
        max_value = (1 << self.active_config.bit_depth) - 1
        return int(clamped * max_value)

    def decode(self, encoded: int) -> float:
        """Decode from active bit depth back to [0,1]"""
        max_value = (1 << self.active_config.bit_depth) - 1
        return encoded / max_value

    def transcode(self, value: int, from_bits: int, to_bits: int) -> int:
        """Convert between different bit depths"""
        from_max = (1 << from_bits) - 1
        to_max = (1 << to_bits) - 1
        normalized = value / from_max
        return int(normalized * to_max)

    @staticmethod
    def select_optimal_bit_depth(task: dict) -> str:
        """Adaptive bit-depth selection"""
        if task.get('realtime'): return 'standard'
        if task.get('iot'): return 'iot_edge'
        if task.get('cryptographic'): return 'rsa_4096'
        if task.get('financial'): return 'octuple'
        if task.get('iterations', 0) > 1e6: return 'octuple'

        accuracy = task.get('accuracy_required', 'medium')
        mapping = {
            'low': 'embedded',
            'medium': 'standard',
            'high': 'high',
            'extreme': 'rsa_4096'
        }
        return mapping.get(accuracy, 'standard')

class MultiResolutionGate:
    """Gates that operate on different bit depths"""

    def __init__(self, config: Union[str, DynamicNibbleConfig] = 'standard'):
        self.nibble_manager = DynamicNibbleManager(config)

    def AND(self, inputs: List[float]) -> float:
        """AND gate with dynamic bit depth"""
        encoded = [self.nibble_manager.encode(v) for v in inputs]
        min_encoded = min(encoded)
        return self.nibble_manager.decode(min_encoded)

    def OR(self, inputs: List[float]) -> float:
        """OR gate with dynamic bit depth"""
        encoded = [self.nibble_manager.encode(v) for v in inputs]
        max_encoded = max(encoded)
        return self.nibble_manager.decode(max_encoded)

    def set_resolution(self, config: Union[str, DynamicNibbleConfig]):
        """Switch resolution at runtime"""
        self.nibble_manager.switch_bit_depth(config)
